src/util/psbt/serialize.rs

```diff
@@ -144,3 +155,178 @@ impl Deserialize for SigHashType {
+        match bytes.len() {
+            65 => {
+                let hash_ty = SchnorrSigHashType::from_u8(bytes[64])
+                    .map_err(|_| encode::Error::ParseFailed("Invalid Sighash type"))?;
+                let sig = schnorrsig::Signature::from_slice(&bytes[..64])
+                    .map_err(|_| encode::Error::ParseFailed("Invalid Schnorr signature"))?;
+                Ok((sig, hash_ty))
+            }
+            64 => {
+                let sig = schnorrsig::Signature::from_slice(&bytes[..64])
+                    .map_err(|_| encode::Error::ParseFailed("Invalid Schnorr signature"))?;
+                Ok((sig, SchnorrSigHashType::Default))
+            }
+            _ => Err(encode::Error::ParseFailed("Invalid Schnorr signature len"))
+        }
+    }
+}
+
+impl Serialize for (schnorr::PublicKey, TapLeafHash) {
+    fn serialize(&self) -> Vec<u8> {
+        let ser_pk = self.0.serialize();
+        let mut buf = Vec::with_capacity(ser_pk.len() + self.1.as_ref().len());
+        buf.extend(&ser_pk);
+        buf.extend(self.1.as_ref());
+        buf
+    }
+}
+
+impl Deserialize for (schnorr::PublicKey, TapLeafHash) {
+    fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error> {
+        if bytes.len() < 32 {
+            return Err(io::Error::from(io::ErrorKind::UnexpectedEof).into())
+        }
+        let a: schnorr::PublicKey = Deserialize::deserialize(&bytes[..32])?;
+        let b: TapLeafHash = Deserialize::deserialize(&bytes[32..])?;
+        Ok((a, b))
+    }
+}
+
+impl Serialize for ControlBlock {
+    fn serialize(&self) -> Vec<u8> {
+        ControlBlock::serialize(&self)
+    }
+}
+
+impl Deserialize for ControlBlock {
+    fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error> {
+        Self::from_slice(bytes)
+            .map_err(|_| encode::Error::ParseFailed("Invalid control block"))
+    }
+}
+
+// Versioned Script
+impl Serialize for (Script, LeafVersion) {
+    fn serialize(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(self.0.len() + 1);
+        buf.extend(self.0.as_bytes());
+        buf.push(self.1.as_u8());
+        buf
+    }
+}
+
+impl Deserialize for (Script, LeafVersion) {
+    fn deserialize(bytes: &[u8]) -> Result<Self, encode::Error> {
+        if bytes.is_empty() {
+            return Err(io::Error::from(io::ErrorKind::UnexpectedEof).into())
+        }
+        // The last byte is LeafVersion.
+        let script = Script::deserialize(&bytes[..bytes.len() - 1])?;
+        let leaf_ver = LeafVersion::from_u8(bytes[bytes.len() - 1])
+            .map_err(|_| encode::Error::ParseFailed("invalid leaf version"))?;
+        Ok((script, leaf_ver))
+    }
+}
+
+
+impl Serialize for (Vec<TapLeafHash>, KeySource) {
+    fn serialize(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity( 32 * self.0.len() + key_source_len(&self.1));
+        self.0.consensus_encode(&mut buf).expect("Vecs don't error allocation");
```

Comment:
Just a note: consensus encoding of Vec len (`usize`) uses `VarInt` which length is different from `self.0.len()` (but I think here it is fine, since the vec can be larger on a few bytes than actually required and we do not risk neither panic or excessive allocation here)